src/high-level/matrix.lisp

```diff
@@ -0,0 +1,288 @@
+                :do (loop :for col :from row :below (ncols matrix)
+                          :do (rotatef
+                               (aref (storage matrix) (funcall index-function (list row col) (shape matrix)))
+                               (aref (storage matrix) (funcall index-function (list col row) (shape matrix))))))
+          (setf (slot-value matrix 'ncols) (first (shape matrix)))
+          (setf (slot-value matrix 'nrows) (second (shape matrix)))))
+    matrix))
+
+(defgeneric diag (matrix)
+  (:documentation "Get a list of the diagonal elements of a matrix")
+  (:method ((matrix matrix))
+    (assert (= (nrows matrix) (ncols matrix))
+            () "Matrix must be square")
+    (let ((rows (nrows matrix)))
+      (loop :for i :below rows
+            :collect (tref matrix i i)))))
+
+(defgeneric inverse (matrix)
+  (:documentation "Get the inverse of the matrix"))
+
+(defgeneric trace (matrix)
+  (:documentation "Get the trace of the matrix (product of diagonals)")
+  (:method ((matrix matrix))
+    (assert (= (nrows matrix) (ncols matrix))
+            () "Matrix must be square")
+    (let ((rows (nrows matrix)))
+      (loop :for i :below rows
+            :sum (tref matrix i i)))))
+
+;; TODO: maybe write generic form of this
+(defgeneric lu (matrix)
+  (:documentation "Get the LU decomposition of the matrix"))
+
+(defgeneric det (matrix)
+  (:documentation "Compute the determinant of a square matrix")
+  (:method ((matrix matrix))
+    (assert (= (nrows matrix) (ncols matrix))
+            () "Matrix must be square")
+    (let ((d 1))
+      (multiple-value-bind (a ipiv) (lu matrix)
+        (dotimes (i (nrows matrix))
+          (setq d (* d (tref a i i))))
+        (dotimes (i (size ipiv))
+          (unless (= (1+ i) (tref ipiv i))
+            (setq d (cl:- d))))
+        d))))
+
+(defgeneric conjugate-transpose (matrix)
+  (:documentation "Compute the conjugate transpose of a matrix")
+  (:method ((matrix matrix))
+    (map #'conjugate matrix)))
+
+(defgeneric conjugate-transpose! (matrix)
+  (:documentation "Compute the conjugate transpose of a matrix, replacing the elements")
+  (:method ((matrix matrix))
+    (map! #'conjugate matrix)))
+
+(defgeneric dagger (matrix)
+  (:documentation "Compute the conjugate transpose of a matrix")
+  (:method ((matrix matrix))
+    (conjugate-transpose matrix)))
+
+(defgeneric dagger! (matrix)
+  (:documentation "Compute the conjugate transpose of a matrix, replacing the elements")
+  (:method ((matrix matrix))
+    (conjugate-transpose! matrix)))
+
+;; TODO: either make this generic or call lapack functions
+(defgeneric orthonormalize! (matrix)
+  (:documentation "Orthonormalize a matrix, replacing the elements"))
+
+;; TODO:
+;; Random unitary
+;; random (different distributions)
+;; diag
+;; Inverse
+;; einsum
+;; Solve
+;; Eigenvalues/vectors
+;; kron
+;; exponent
+;; conj-transpose
+;; csd
+;; svd
+;; ql
+;; qr
+;; rq
+;; lq
+;; dot
+;; get diag elems?
+;; stack together
+
```

Comment:
Once you have the SVD, you could provide a function to get the two-norm of the matrix (i.e., the largest singular value).